src/source.rs

```diff
@@ -181,35 +186,297 @@ where
+    }
+
+    /// Get the maximum size of an encoded and decoded gRPC message in bytes. Default value is 64MB.
+    pub fn max_message_size(&self) -> usize {
+        self.max_message_size
+    }
+
+    /// Change the file in which numflow server information is stored on start up to the new value. Default value is `/tmp/numaflow.server-info`
+    pub fn with_server_info_file(mut self, file: impl Into<PathBuf>) -> Self {
+        self.server_info_file = file.into();
+        self
+    }
+
+    /// Get the path to the file where numaflow server info is stored. Default value is `/tmp/numaflow.server-info`
+    pub fn server_info_file(&self) -> &std::path::Path {
+        self.server_info_file.as_path()
+    }
+
+    /// Starts the gRPC server. When message is received on the `shutdown` channel, graceful shutdown of the gRPC server will be initiated.
+    pub async fn start_with_shutdown(
+        &mut self,
+        shutdown: oneshot::Receiver<()>,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>>
+    where
+        T: Sourcer + Send + Sync + 'static,
+    {
+        let listener = shared::create_listener_stream(&self.sock_addr, &self.server_info_file)?;
+        let handler = self.svc.take().unwrap();
+        let source_service = SourceService {
+            handler: Arc::new(handler),
+        };
+
+        let source_svc = source_server::SourceServer::new(source_service)
+            .max_decoding_message_size(self.max_message_size)
+            .max_decoding_message_size(self.max_message_size);
+
+        let shutdown = async {
+            shutdown
+                .await
+                .expect("Receiving message from shutdown channel");
+        };
+        tonic::transport::Server::builder()
+            .add_service(source_svc)
+            .serve_with_incoming_shutdown(listener, shutdown)
+            .await
+            .map_err(Into::into)
+    }
+
+    /// Starts the gRPC server. Automatically registers singal handlers for SIGINT and SIGTERM and initiates graceful shutdown of gRPC server when either one of the singal arrives.
+    pub async fn start(&mut self) -> Result<(), Box<dyn std::error::Error + Send + Sync>>
+    where
+        T: Sourcer + Send + Sync + 'static,
+    {
+        let (tx, rx) = oneshot::channel::<()>();
+        tokio::spawn(shared::wait_for_signal(tx));
+        self.start_with_shutdown(rx).await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::sourcer;
+    use chrono::Utc;
+    use std::vec;
+    use std::{error::Error, time::Duration};
+    use tokio::sync::Mutex;
+    use tokio_stream::StreamExt;
+    use tower::service_fn;
+
+    use crate::source::{self, Message, Offset, SourceReadRequest};
+    use tempfile::TempDir;
+    use tokio::sync::mpsc::Sender;
+    use tokio::sync::oneshot;
+    use tonic::transport::Uri;
+
+    // A source that returns numbers in the range [0..end)
+    struct Counter {
+        end: usize,
+        pos: Mutex<usize>,
+        acked: Mutex<usize>,
```

Comment:
do these need to be tokio::Mutex?

Reply:
`pos` is held across `.await` points in `read` function. So `tokio::Mutex` is needed. `acked` can use `std::sync::Mutex`